[클린코드 8기 박래욱] 로또 미션 STEP1

ricolazerosugar opened the pull request:
## 구현 기능
로또 기능 STEP1

## 하고싶은 말
- 전반적으로 구현하는데 급급해서 아쉬움이 많습니다.
- 더 혼자 개선하기 보다는 피드백을 우선 받아보려고 합니다.
- 피드백에 열려 있겠습니다. 잘 부탁드려요.

giraffeb committed:
chore: 콘솔 로또 요구사항 작성

giraffeb committed:
feat: 입력 UI 추가 - 로또 구입 금액, 당첨 번호, 보너스 번호

giraffeb committed:
feat: 로또 금액 1000원 설정

giraffeb committed:
feat: 로또 구입 금액에 해당하는 만큼 로또를 발행

giraffeb committed:
feat: 로또 번호는 1부터 45까지 번호 중 중복 없이 6개를 생성한다.

giraffeb committed:
feat: 당첨 번호와 보너스 번호를 저장한다.

giraffeb committed:
feat: 사용자가 구매한 로또 번호와 당첨 번호를 비교하여 당첨 내역을 계산한다.

giraffeb committed:
feat: 사용자의 당첨내역의 수익률을 계산한다.

giraffeb committed:
fix: 로또 번호 생성이 랜덤으로 처리되지 않는 문제

giraffeb committed:
feat: 당첨 내역과 수익률을 출력할 수 있게 한다.

- 사용자가 구매한 로또 번호와 당첨 번호를 비교하여 당첨 내역 및 수익률을 출력한다.

giraffeb committed:
fix: 변수명 오타

giraffeb committed:
refact: 로또와 보너스 로또 공통 로직 통합

giraffeb committed:
feat: 구입금액 에러 처리 추가

yujo11 commented on src/lotto/lotto.js:
```diff
@@ -0,0 +1,184 @@
+    return this.#numbers;
+  }
+
+  _validateLottoNumbers(lottoNumbers) {
```
이 `_`(언더바)는 어떤 의미로 사용하신건지 궁금합니다 !

ricolazerosugar replied:
private 힌 메소드를 의미한다고 작성했는데, 메소드에도 #를 붙일수 있군요! ...

https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Classes/Private_elements

yujo11 replied:
맞습니다! 언어에서 제공하는 스펙을 따라가면 좋을거 같아요 ㅎㅎ 👍 👍 👍

yujo11 commented on src/lotto/lotto.js:
```diff
@@ -0,0 +1,184 @@
+}
+
+class LottoWinningRule {
+  static FIRST_PRIZE = {
```
상수로 사용할 수 있도록 정의해주셨군요 👍 👍 👍

yujo11 commented on src/lotto/lotto.js:
```diff
@@ -0,0 +1,184 @@
+  static FIRST_PRIZE = {
+    matchedNumberCount: 6,
+    matchedBonusNumberCount: 0,
+    prize: 2000000000,
```
큰 숫자를 사용할 때는 값을 쉽게 인지할 수 있도록 numeric seprator를 사용해보셔도 좋을거 같아요 !
```suggestion
    prize: 2_000_000_000,
```

yujo11 commented on src/lotto/lotto.js:
```diff
@@ -0,0 +1,184 @@
+  }
+}
+
+class LottoWinningRule {
```
모든 값, 메서드가 `static`으로 선언되어 있는데요.  `LottoWinningRule`이 class가 아니어도 되었을거 같은데 class로 만드신 이유가 궁금합니다!

ricolazerosugar replied:
class로 모듈화를 하고 싶었어요. 의견 보고  생각해보니 함수로만 만들고 모듈화를 class로 하지 않는게 더 직관적일 수 있겠네요.

yujo11 commented on src/step1-index.js:
```diff
@@ -2,3 +2,69 @@
+  console.log(`총 수익률은 ${winningReport.winningRate.toFixed(1)}%입니다.`);
+};
+
+const main = async () => {
```
여기에 입력 처리, 검증, 로직 실행 등 많은 코드들이 작성되어 있는데요. 각 기능별로 함수를 나눠보는건 어떨까요?

yujo11 reviewed (changes requested):
안녕하세요 래욱님~~ 만나서 반갑습니다! 로또 미션 잘 부탁드립니다!!

>더 혼자 개선하기 보다는 피드백을 우선 받아보려고 합니다.

아주 좋은 접근 방법이라고 생각합니다! 혼자서만 진행하는 과정이 아닌만큼 리뷰어와 코치 분들을 적극적으로 이용(?)하시면 좋을거 같아요!

진행하시면서 고민되는 부분들이 있었다면 리뷰요청 주실 떄 같이 남겨주시면 더 좋을거 같습니다 ㅎㅎ 👍 👍 👍 

---

전반적으로 소소한 코멘트와 코드를 보며 궁금한 점들에 질문을 남겼는데요. 확인해주시면 감사하겠습니다!

미션 진행하느라 고생 많으셨습니다 ~~ 👍 👍 👍

giraffeb committed:
refact: 로또 당첨금 numeric separator 적용

giraffeb committed:
refact: 로또 private method 문법 적용

giraffeb committed:
refact: Lotto 상수 개선

giraffeb committed:
feat: 로또 UI 컴포넌트 변경

giraffeb committed:
feat: 로또 구매 및 당첨 결과 처리 로직 추가

- 로또 구매를 위한 LottoStore 클래스 추가
- 사용자가 구매한 로또 번호와 당첨 번호를 비교하여 당첨 개수 및 보너스 번호 당첨 여부를 반환하는 기능 구현
- 당첨 개수에 따라 당첨 등수를 반환하는 로직 추가
- 콘솔 UI를 통해 로또 구매 및 결과 출력 기능 구현

giraffeb committed:
fix: 로또 당첨 계산 이름 변경 및 로직 분리

ricolazerosugar commented:
안녕하세요. 윤호님!
오랜만에 리뷰요청드리네요

피드백 주신 입력 처리, 검증, 로직 실행을 분리해 보았습니다. 다른 부분도 일부 개선해 보았습니다.

잘 부탁드립니다.

yujo11 commented on src/lotto/lotto.js:
```diff
@@ -91,12 +91,12 @@ class LottoWinningRule {
     matchedNumberCount: 6,
     matchedBonusNumberCount: 0,
-    prize: 2000000000,
+    prize: 2_000_000_000,
```
👍 👍 👍

yujo11 commented on src/lotto/lotto.js:
```diff
@@ -15,7 +15,7 @@ class CommonLotto {
   }
 
-  _validateLottoNumbers(lottoNumbers) {
+  #validateLottoNumbers(lottoNumbers) {
```
👍 👍 👍

yujo11 commented on src/step1-index.js:
```diff
@@ -24,13 +24,37 @@ const printLottoReport = (winningReport) => {
+  lottos,
+  lottoNumberString,
+  bonusNumberString
+) => {
```
여러개의 파라미터를 받을 때는 객체로 파라미터를 받는 방법도 고려해보시면 좋을거 같아요 ㅎㅎ

yujo11 commented on src/lotto/lotto.js:
```diff
@@ -27,7 +27,7 @@ class CommonLotto {
-    return [...LOTTO_NUMBER_RANGE]
+    return [...CommonLotto.LOTTO_NUMBER_RANGE]
       .sort(() => Math.random() - 0.5)
       .splice(0, numberCount);
```
`splice()`와 `slice()` 메서드는 어떤 차이가 있을까요? 그리고 객체의 불변성을 지키기 위해서는 어떤 메서드를 사용하는게 더 유리할까요?

yujo11 reviewed (approved):
래욱님 안녕하세요 ~~ 오래만에 뵙네요 ㅎㅎ

소소한 코멘트 추가로 남겼으니 확인해보시고 다음 미션 진행할 때 반영해보시면 좋을거 같아요!

고생 많으셨습니다 ~~ 👍 👍 👍
